benchmarks/benchmarks/generate_data/um_files.py

```diff
@@ -0,0 +1,212 @@
+# Copyright Iris contributors
+#
+# This file is part of Iris and is released under the LGPL license.
+# See COPYING and COPYING.LESSER in the root of the repository for full
+# licensing details.
+"""
+Generate FF, PP and NetCDF files based on a minimal synthetic FF file.
+
+NOTE: uses the Mule package, so depends on an environment with Mule installed.
+"""
+
+
+def _create_um_files(
+    len_x: int, len_y: int, len_z: int, len_t: int, compress, save_paths: dict
+) -> None:
+    """
+    Generate an FF object of given shape and compression, save to FF/PP/NetCDF.
+
+    This is run externally
+    (:func:`benchmarks.generate_data.run_function_elsewhere`), so all imports
+    are self-contained and input parameters are simple types.
+    """
+    from copy import deepcopy
+    from datetime import datetime
+    from tempfile import NamedTemporaryFile
+
+    from mo_pack import compress_wgdos as mo_pack_compress
+    from mule import ArrayDataProvider, Field3, FieldsFile
+    from mule.pp import fields_to_pp_file
+    import numpy as np
+
+    from iris import load_cube
+    from iris import save as save_cube
+
+    def packing_patch(*compress_args, **compress_kwargs) -> bytes:
+        """
+        Force conversion from returned :class:`memoryview` to :class:`bytes`.
+
+        Downstream uses of :func:`mo_pack.compress_wgdos` were written
+        for the ``Python2`` behaviour, where the returned buffer had a
+        different ``__len__`` value to the current :class:`memoryview`.
+        Unable to fix directly in Mule, so monkey patching for now.
+        """
+        return mo_pack_compress(*compress_args, **compress_kwargs).tobytes()
+
+    import mo_pack
+
+    mo_pack.compress_wgdos = packing_patch
+
+    ########
+
+    template = {
+        "fixed_length_header": {"dataset_type": 3, "grid_staggering": 3},
+        "integer_constants": {
+            "num_p_levels": len_z,
+            "num_cols": len_x,
+            "num_rows": len_y,
+        },
+        "real_constants": {},
+        "level_dependent_constants": {"dims": (len_z + 1, None)},
+    }
+    new_ff = FieldsFile.from_template(deepcopy(template))
+
+    data_array = np.arange(len_x * len_y).reshape(len_x, len_y)
+    array_provider = ArrayDataProvider(data_array)
+
+    def add_field(level_: int, time_step_: int) -> None:
```

Comment:
I don't understand why the arg names `level_`, `time_step_`  have the final underscores.
Does this have a particular meaning that I'm missing ?

Reply:
Some IDE's (I'm using PyCharm) warn about 'shadowing outer scope' - where variables have the same name as the next level up.

I assume this is from the risk of confusion where one could delete the 'inner' variable and the script might still run fine, just not in the way intended.

Reply:
Ok thanks, wasn't really aware of that.
Interesting : I'm sure I do write code that might trigger this -- so, my PyCharm had that warning turned off for some reason.
Anyway it's a reasonable approach to avoid those warnings. :+1: